File: infra-document/src/main/java/top/emilejones/hhu/document/config/MybatisConfig.java
package top.emilejones.hhu.document.config;

import org.apache.ibatis.session.SqlSessionFactory;
import org.mybatis.spring.SqlSessionFactoryBean;
import org.mybatis.spring.SqlSessionTemplate;
import org.mybatis.spring.annotation.MapperScan;
import org.springframework.beans.factory.annotation.Qualifier;
import org.springframework.boot.jdbc.DataSourceBuilder;
import org.springframework.context.annotation.Bean;
import org.springframework.context.annotation.Configuration;
import org.springframework.core.io.support.PathMatchingResourcePatternResolver;
import top.emilejones.hhu.env.pojo.MysqlConfig;

import javax.sql.DataSource;

/**
 * MyBatis wiring that binds mapper scanning to the connection beans provided by MysqlConnectionConfig.
 */
@Configuration
@MapperScan(
        basePackages = "top.emilejones.hhu.document.mapper",
        sqlSessionFactoryRef = "infra-document-SqlSessionFactory"
)
public class MybatisConfig {

    @Bean(name = "infra-document-DataSource")
    public DataSource dataSource(MysqlConfig mysqlConfig) {
        String url = String.format(
                "jdbc:mysql://%s:%d/%s?useSSL=false&characterEncoding=utf8&serverTimezone=UTC",
                mysqlConfig.getHost(),
                mysqlConfig.getPort(),
                mysqlConfig.getDatabase()
        );
        return DataSourceBuilder.create()
                .driverClassName("com.mysql.cj.jdbc.Driver")
                .url(url)
                .username(mysqlConfig.getUser())
                .password(mysqlConfig.getPassword())
                .build();
    }

    @Bean(name = "infra-document-SqlSessionFactory")
    public SqlSessionFactory sqlSessionFactory(DataSource dataSource) throws Exception {
        SqlSessionFactoryBean factory = new SqlSessionFactoryBean();
        factory.setDataSource(dataSource);
        factory.setMapperLocations(
                new PathMatchingResourcePatternResolver().getResources("classpath*:mapper/*.xml")
        );
        return factory.getObject();
    }

    @Bean
    public SqlSessionTemplate sqlSessionTemplate(@Qualifier("infra-document-SqlSessionFactory") SqlSessionFactory sqlSessionFactory) {
        return new SqlSessionTemplate(sqlSessionFactory);
    }
}

File: infrastructure/environment/src/main/kotlin/top/emilejones/hhu/infrastructure/configuration/beans/MybatisConfig.kt
package top.emilejones.hhu.infrastructure.configuration.beans

import org.apache.ibatis.session.SqlSessionFactory
import org.mybatis.spring.SqlSessionFactoryBean
import org.mybatis.spring.SqlSessionTemplate
import org.mybatis.spring.annotation.MapperScan
import org.springframework.beans.factory.annotation.Qualifier
import org.springframework.boot.jdbc.DataSourceBuilder
import org.springframework.context.annotation.Bean
import org.springframework.context.annotation.Configuration
import org.springframework.context.annotation.Primary
import org.springframework.core.io.support.PathMatchingResourcePatternResolver
import org.springframework.jdbc.datasource.DataSourceTransactionManager
import top.emilejones.hhu.infrastructure.configuration.env.pojo.MysqlConfig
import javax.sql.DataSource

@Configuration
@MapperScan(basePackages = ["top.emilejones.hhu.**.mapper"])
class MybatisConfig {

    @Bean
    @Primary
    fun dataSource(mysqlConfig: MysqlConfig): DataSource {
        val url = "jdbc:mysql://%s:%d/%s?useUnicode=true&characterEncoding=utf-8&useSSL=false&allowPublicKeyRetrieval=true".format(
            mysqlConfig.host,
            mysqlConfig.port,
            mysqlConfig.database
        )
        return DataSourceBuilder.create()
            .driverClassName("com.mysql.cj.jdbc.Driver")
            .url(url)
            .username(mysqlConfig.user)
            .password(mysqlConfig.password)
            .build()
    }

    @Bean
    @Primary
    fun sqlSessionFactory(@Qualifier("dataSource") dataSource: DataSource): SqlSessionFactory {
        val factoryBean = SqlSessionFactoryBean()
        factoryBean.setDataSource(dataSource)
        factoryBean.setMapperLocations(
            *PathMatchingResourcePatternResolver().getResources("classpath*:mapper/*.xml")
        )
        // Set type handlers package to a common root or specific packages if known
        // Assuming scanning from root package recursively for type handlers
        factoryBean.setTypeHandlersPackage("top.emilejones.hhu.knowledge.handler,top.emilejones.hhu.common.handler") 
        return factoryBean.`object`!!
    }

    @Bean
    @Primary
    fun sqlSessionTemplate(@Qualifier("sqlSessionFactory") sqlSessionFactory: SqlSessionFactory): SqlSessionTemplate {
        return SqlSessionTemplate(sqlSessionFactory)
    }

    @Bean
    @Primary
    fun transactionManager(@Qualifier("dataSource") dataSource: DataSource): DataSourceTransactionManager {
        return DataSourceTransactionManager(dataSource)
    }
}
